File: Problems/1161. Maximum Level Sum of a Binary Tree/Maximum_Level_Sum_of_a_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxLevelSum(TreeNode* root) {
        if (root == nullptr){
            return 0;
        }
        std::queue<TreeNode*> q;
        q.push(root);
        int current_level{1}, current_sum{0};
        int highest_sum_level{1}, highest_sum{-999999};

        while (q.size() > 0){
            size_t queue_length = q.size();
            for (int i = 0; i < queue_length; i ++){
                TreeNode* current = q.front();
                q.pop();
                if (current != nullptr){
                    // Append children
                    if (current->left){ q.push(current->left);}
                    if (current->right){ q.push(current->right);}

                    current_sum += current->val;
                }
            }
            if (current_sum > highest_sum) {
                highest_sum = current_sum;
                highest_sum_level = current_level;
            }
            current_sum = 0;
            current_level++;
        }
        return highest_sum_level;
    }
};

File: Problems/1161. Maximum Level Sum of a Binary Tree/Maximum_Level_Sum_of_a_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def maxLevelSum(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        if not root: 
            return None
        # Use queue to track nodes per level
        queue = deque()
        queue.append(root)
        level_total, current_level = 0, 1
        highest_sum, highest_sum_level = -999999999, 1
        while queue:
            # Pop all from queue and keep track of sum. Order doesn't matter
            for i in range(len(queue)):
                popped_node = queue.popleft()
                if popped_node:
                    level_total += popped_node.val
                    # Transfer children if they are defined
                    if popped_node.left: queue.append(popped_node.left)
                    if popped_node.right: queue.append(popped_node.right)

            if level_total > highest_sum:
                highest_sum = level_total
                highest_sum_level = current_level
            level_total = 0
            # Swap queue, reset children, and increment level
            current_level += 1
        return highest_sum_level

        